Support threading for zstd compression.

marxin opened the pull request:
Port zstd compression to the new API (ZSTD_compressStream2).

My motivation is to eventually enable multi-threaded compression which can help packages like Firefox where there's a single huge rpm that is being compressed.

Once the change is accepted, the following one-liner will enable parallel compression:
```c
ZSTD_CCtx_setParameter(cctx, ZSTD_c_nbWorkers, 4);
```

marxin commented:
Demo of the used parallel compression for `cvise` package:
https://gist.github.com/marxin/5560256dbf0af4c749bd4d56c0e83626

mlschroe commented:
You're aware that this breaks delta rpms?

marxin commented:
No. How does it break it?

mlschroe commented:
You need to recreate the same bytewise identical compression when applying the deltarpm. The parallel compression implementations tend to partition the data and compress the chunks individually, leading to a different result.

I haven't checked the zstd implementation, I admit, I just know that this was the problem with parallel bzip2/xz compression.

marxin commented:
> I haven't checked the zstd implementation, I admit, I just know that this was the problem with parallel bzip2/xz compression.

I've made some experiments for a 100MB big file and the compressed output is still the same.
For being sure, I asked the question in https://github.com/facebook/zstd/issues/2238.

marxin commented:
So the answer is that it's stable and reproducible.

ffesti commented:
This change looks good and straight forward.

I am missing some information on what version of zstd supports the new API - and if necessary a check in configure.ac.

marxin commented:
> This change looks good and straight forward.

Thanks!

> 
> I am missing some information on what version of zstd supports the new API - and if necessary a check in configure.ac.

Updated `configure.ac` (one needs at least release `1.3.8`).

ignatenkobrain commented on configure.ac:
```diff
@@ -184,13 +184,27 @@ AC_ARG_ENABLE([zstd],
       AC_MSG_ERROR([--enable-zstd specified, but not available])
     fi
   ])
+  PKG_CHECK_MODULES([ZSTD], [libzstd], [have_zstd=yes], [have_zstd=no])
```
I think it is much easier to have here `libzstd >= 1.3.8` rather than custom C preproc check.

marxin replied:
Oh yeah, it is new to me! Thanks.

ignatenkobrain commented on rpmio/rpmio.c:
```diff
@@ -1134,10 +1139,15 @@ static rpmzstd rpmzstdNew(int fdno, const char *fmode)
 	    goto err;
 	}
+
+	if (threads > 0)
```
If I read https://facebook.github.io/zstd/zstd_manual.html correctly, even if you set it to `0`, it will use single thread which makes this if unneeded.

ignatenkobrain replied:
hmmm, thinking about it more..

```
    /* These parameters are only useful if multi-threading is enabled (compiled with build macro ZSTD_MULTITHREAD).
```

And checking code, yeah - it will return and error. so should we instead change it to `> 1`?

marxin replied:
Good point! I changed that to `> 1`, same what lzma does.

ignatenkobrain commented on rpmio/rpmio.c:
```diff
@@ -1100,7 +1101,12 @@ static rpmzstd rpmzstdNew(int fdno, const char *fmode)
 	    flags |= O_RDWR;
 	    continue;
-	    break;
+	case 'T':
```
I think it would be nice to support setting it to -1, same as for lzma:

```
                if (threads == -1)
                    threads = rpmExpandNumeric("%{getncpus}");
```

ignatenkobrain commented on rpmio/rpmio.c:
```diff
@@ -1100,7 +1101,12 @@ static rpmzstd rpmzstdNew(int fdno, const char *fmode)
+	case 'T':
+	    ++s;
+	    c = *s;
+	    if (c >= (int)'0' && c <= (int)'9')
+		threads = strtol(s-1, (char **)&s, 10);
```
Just curious if it would make sense to change lzopen_internal() to follow the same logic? If so, could you push a separate commit for that?

marxin replied:
Well, it's just a small piece of code. I would not put it to a separate function.

ignatenkobrain commented:
@marxin I've tried this on my laptop and it does not seem to work?

```
$ /usr/bin/time ~/Projects/upstream/rpm/rpmbuild -bb xonotic-data.spec -D "_sourcedir $PWD" -D "_binary_payload w19T8.zstdio"
...
289.90user 3.08system 4:54.67elapsed 99%CPU (0avgtext+0avgdata 155348maxresident)k
0inputs+5456184outputs (0major+72889minor)pagefaults 0swaps
```

ignatenkobrain commented on rpmio/rpmio.c:
```diff
@@ -1100,7 +1101,12 @@ static rpmzstd rpmzstdNew(int fdno, const char *fmode)
 	    continue;
-	    break;
+	case 'T':
+	    ++s;
+	    c = *s;
```
```suggestion
	    c = *s++;
```

ignatenkobrain commented on rpmio/rpmio.c:
```diff
@@ -1134,10 +1139,15 @@ static rpmzstd rpmzstdNew(int fdno, const char *fmode)
+
+	if (threads > 0)
+	  if (ZSTD_isError (ZSTD_CCtx_setParameter(_stream, ZSTD_c_nbWorkers, threads)))
+	    rpmlog(RPMLOG_WARNING, "zstd library does not support multi-threading");
```
```suggestion
	    rpmlog(RPMLOG_WARNING, "zstd library does not support multi-threading\n");
```

ignatenkobrain commented on rpmio/rpmio.c:
```diff
@@ -1134,10 +1139,15 @@ static rpmzstd rpmzstdNew(int fdno, const char *fmode)
 	}
+
+	if (threads > 0)
+	  if (ZSTD_isError (ZSTD_CCtx_setParameter(_stream, ZSTD_c_nbWorkers, threads)))
```
hmmpf, this always gives me an error on Fedora Rawhide:

```
warning: zstd library does not support multi-threading
```

With `ZSTD_getErrorName()` I get `Unsupported parameter`.

ignatenkobrain replied:
Oh my, Fedora ships libzstd without MT support. ignore this one.

ignatenkobrain commented on rpmio/rpmio.c:
```diff
@@ -1134,10 +1139,15 @@ static rpmzstd rpmzstdNew(int fdno, const char *fmode)
 	    goto err;
 	}
+
+	if (threads > 0)
```
```suggestion
	if (threads == -1)
	    threads = rpmExpandNumeric("%{getncpus}");
	if (threads > 0)
```

ignatenkobrain commented:
With my suggestions, and a proper build of libzstd in Fedora it works as expected, it fully utilizes my system. Testing on `xonotic-data` which is 873M.

```
before: 
284.05user 3.03system 4:48.64elapsed 99%CPU (0avgtext+0avgdata 153540maxresident)k
88inputs+5456312outputs (1major+80373minor)pagefaults 0swaps

after:
578.54user 5.56system 1:36.86elapsed 602%CPU (0avgtext+0avgdata 1409788maxresident)k
104inputs+5455928outputs (1major+386345minor)pagefaults 0swaps
```

marxin commented:
> With my suggestions, and a proper build of libzstd in Fedora it works as expected, it fully utilizes my system.

Great to hear!

ignatenkobrain commented on rpmio/rpmio.c:
```diff
@@ -1134,10 +1138,17 @@ static rpmzstd rpmzstdNew(int fdno, const char *fmode)
+
+	if (threads == -1)
+	    threads = rpmExpandNumeric("%{getncpus}");
+	if (threads > 1)
```
After all I think

```suggestion
	if (threads > 0)
```

should be used.

mlschroe commented:
Note that the zstd people *did* say that threaded compression has a different output than unthreaded operation: it compresses a bit worse. So the pull request will break delta rpms.

Can you please change the code so that it uses unthreaded mode if there is no 'T' in the compression flags? I.e. initialize `threads` with 0 instead of -1?

mlschroe commented:
I'm also not so sure about the "the multi-threaded output produces the same compressed data no matter how many threads you use" statement, because the block size seems to depend on the number of workers (see ZSTDMT_compress_advanced_internal).

ignatenkobrain commented:
> it compresses a bit worse

I was testing it on xonotic-data and it was 873M in single-threaded compression and the same size in multi-threaded mode.

mlschroe commented:
But was it identical? Zstd upstream said that it is different.

mlschroe commented:
Also note that threading mode with one thread (T1) is *not* unthreaded mode.

mlschroe commented:
Hmm, I don't think ZSTDMT_compress_advanced_internal is used in out use case, so disregard the comment about the influence of the thread number. (I'm not 100% sure, though, the zstd source code is somewhat cryptic to me.)

mlschroe commented:
$ rpm2cpio kdelibs3-3.5.5-39.i586.rpm | zstd -c --single-thread | wc -c
16307922
$ rpm2cpio kdelibs3-3.5.5-39.i586.rpm | zstd -c -T1 | wc -c
16308523

ffesti commented:
We need to make sure that the result with the current settings (non threaded compression) remains unchanged to not break deltarpm that way.

ignatenkobrain commented on rpmio/rpmio.c:
```diff
@@ -1072,6 +1072,7 @@ static rpmzstd rpmzstdNew(int fdno, const char *fmode)
     char *t = stdio;
     char *te = t + sizeof(stdio) - 2;
     int c;
+    int threads = -1;
```
```suggestion
    int threads = 0;
```

to satisfy @ffesti

marxin commented:
All right, I changed the logic so that threaded mode is used only when a user set a `T` value in the compression algorithm.
And I've just removed the automatic deduction of number of cores for now.
@mlschroe Hope it's useful now?

marxin commented:
> I'm also not so sure about the "the multi-threaded output produces the same compressed data no matter how many threads you use" statement, because the block size seems to depend on the number of workers (see ZSTDMT_compress_advanced_internal).

I've just asked about this explicitly here:
https://github.com/facebook/zstd/issues/2238#issuecomment-662906821

marxin commented:
> From my POV it is good to go and we can follow up with improvements like auto-detection of cores.

Thank you.
Yes, I agree with that. I can imagine doing what `zstd` or `xz` do for auto-detection of cores: `xz -T0 `.

marxin commented:
> I've just asked about this explicitly here:
> [facebook/zstd#2238 (comment)](https://github.com/facebook/zstd/issues/2238#issuecomment-662906821)

And `zstd` folks confirmed that a number of threads does not influence the stability of the "threaded" compression mode.

mlschroe commented:
Looks good to me, thanks. (Of course the deltarpm and drpm need to be updated now so that they support a ZSTD_THREADED compression type as well.)

marxin commented:
> Looks good to me, thanks. (Of course the deltarpm and drpm need to be updated now so that they support a ZSTD_THREADED compression type as well.)

Thanks.

I can start working on the support for the mentioned tools. Can you please guide me a bit what would be needed?

mlschroe commented:
I'll implement support in deltarpm and you can port it to drpm, ok?

marxin commented:
> I'll implement support in deltarpm and you can port it to drpm, ok?

Sure, appreciate that!

Conan-Kudo commented on macros.in:
```diff
@@ -350,7 +350,7 @@ package or when debugging this package.\
 #		"w9.bzdio"	bzip2 level 9.
 #		"w6.xzdio"	xz level 6, xz's default.
-#		"w7T16.xzdio"	xz level 7 using 16 thread (xz only)
+#		"w7T16.xzdio"	xz level 7 using 16 thread (xz and zstd only)
```
This description is wrong. You need a new line describing this for zstd.

marxin replied:
I understand it that the description in parenthesis is related to the threading mode.
And `zstd` newly support it. Or am I wrong?

Conan-Kudo replied:
`zstd` would use `w7T16.zstdio`, so it needs its own line.

marxin commented:
Thank you for the review. Having 2 approvals can please anybody merge it?

ffesti commented:
Looks like there's a hick-up in the CI. Re-running the semaphore tests.

marxin commented:
> Looks like there's a hick-up in the CI. Re-running the semaphore tests.
> 
> Well, trying...

Thanks, apparently it's still stuck..

marxin commented:
Can please anybody unblock it?

ffesti commented:
Looks like it fails to detect the right commit to checkout and build. Can you force push a new version, please, to test if that get's it unstuck? Just do some white space changes to the commit message.

marxin commented:
> Looks like it fails to detect the right commit to checkout and build. Can you force push a new version, please, to test if that get's it unstuck? Just do some white space changes to the commit message.

Done, good point!

marxin commented:
CI seems happy now, can please anybody merge it? I've got another patch based on this one that I would like to propose..

ffesti commented:
Sorry, this took a bit longer than it should have, Thanks for your work and patience!
